rust/catalyst-types/src/mmap_file.rs

```diff
@@ -0,0 +1,193 @@
+//! Memory-mapped file.
+
+use core::fmt;
+use std::{
+    path::Path,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc,
+    },
+};
+
+use fmmap::{MmapFile, MmapFileExt};
+use once_cell::sync::Lazy;
+use serde::{ser::SerializeStruct, Serialize, Serializer};
+
+/// Memory-mapped file.
+pub struct MemoryMapFile {
+    /// The memory-mapped file.
+    file: MmapFile,
+    /// The size of the memory-mapped file.
+    size: u64,
+}
+/// Global statistic for memory-mapped files.
+static MEMMAP_FILE_STATS: Lazy<MemMapFileStat> = Lazy::new(MemMapFileStat::default);
+
+/// Memory-mapped file statistic.
+#[derive(Debug, Default, Clone, Serialize)]
+pub struct MemMapFileStat(Arc<MemMapFileStatInner>);
+
+/// Internal structure to hold stats.
+struct MemMapFileStatInner {
+    /// A counter for the number of memory-mapped files.
+    file_count: AtomicU64,
+    /// The total size of memory-mapped files.
+    total_size: AtomicU64,
+    /// The amount of time that memory-mapped files have been dropped.
+    drop_count: AtomicU64,
+    /// The total size of memory-mapped files that have been dropped.
+    drop_size: AtomicU64,
+    /// A count of errors encountered.
+    error_count: AtomicU64,
+}
+
+impl MemMapFileStat {
+    /// Get the statistic file count.
+    #[must_use]
+    pub fn file_count(&self) -> u64 {
+        self.0.file_count.load(Ordering::SeqCst)
+    }
+
+    /// Get the statistic total size.
+    #[must_use]
+    pub fn total_size(&self) -> u64 {
+        self.0.total_size.load(Ordering::SeqCst)
+    }
+
+    /// Get the statistic drop count.
+    #[must_use]
+    pub fn drop_count(&self) -> u64 {
+        self.0.drop_count.load(Ordering::SeqCst)
+    }
+
+    /// Get the statistic drop size.
+    #[must_use]
+    pub fn drop_size(&self) -> u64 {
+        self.0.drop_size.load(Ordering::SeqCst)
+    }
+
+    /// Get the statistic error count.
+    #[must_use]
+    pub fn error_count(&self) -> u64 {
+        self.0.error_count.load(Ordering::SeqCst)
+    }
+}
+
+impl MemoryMapFile {
+    /// Get the memory-mapped file.
+    pub fn file(&self) -> &MmapFile {
+        &self.file
+    }
+
+    /// Get the memory-mapped file as a slice.
+    pub fn file_as_slice(&self) -> &[u8] {
+        self.file().as_slice()
+    }
+
+    /// Get the size of the memory-mapped file.
+    pub fn size(&self) -> u64 {
+        self.size
+    }
+
+    /// Get the global memory-mapped file statistics.
+    #[must_use]
+    pub fn stat() -> &'static MemMapFileStat {
+        &MEMMAP_FILE_STATS
```

Comment:
Return a clone of the underlying stats, so they don't change after this call asynchronously.
Not a clone of the Arc()